dask/dataframe/io/io.py

```diff
@@ -454,48 +454,59 @@ def from_dask_array(x, columns=None, index=None, meta=None):
     """
     meta = _meta_from_array(x, columns, index, meta=meta)
 
-    if x.ndim == 2 and len(x.chunks[1]) > 1:
-        x = x.rechunk({1: x.shape[1]})
+    name = "from-dask-array-" + tokenize(x, columns)
+    graph_dependencies = [x]
+    arrays_and_indices = [x.name, "ij" if x.ndim == 2 else "i"]
+    numblocks = {x.name: x.numblocks}
 
-    name = "from-dask-array" + tokenize(x, columns)
-    to_merge = []
+    if is_series_like(meta):
+        kwargs = {"dtype": x.dtype, "name": meta.name}
+    else:
+        kwargs = {"columns": meta.columns}
 
     if index is not None:
-        if not isinstance(index, Index):
-            raise ValueError("'index' must be an instance of dask.dataframe.Index")
         if index.npartitions != x.numblocks[0]:
             msg = (
                 "The index and array have different numbers of blocks. "
                 "({} != {})".format(index.npartitions, x.numblocks[0])
             )
             raise ValueError(msg)
         divisions = index.divisions
-        to_merge.append(ensure_dict(index.dask))
-        index = index.__dask_keys__()
+        graph_dependencies.append(index)
+        arrays_and_indices.extend([index.__dask_layers__()[0], "i"])
+        numblocks[index.__dask_layers__()[0]] = (index.npartitions,)
 
     elif np.isnan(sum(x.shape)):
         divisions = [None] * (len(x.chunks[0]) + 1)
-        index = [None] * len(x.chunks[0])
     else:
+        dsk = {}
         divisions = [0]
-        for c in x.chunks[0]:
-            divisions.append(divisions[-1] + c)
-        index = [
-            (np.arange, a, b, 1, "i8") for a, b in zip(divisions[:-1], divisions[1:])
-        ]
+        stop = 0
+        index_name = name + "-index-" + tokenize(x.chunks[0])
+        for i, increment in enumerate(x.chunks[0]):
+            stop += increment
+            dsk[(index_name, i)] = (pd.RangeIndex, divisions[i], stop)
+            divisions.append(stop)
         divisions[-1] -= 1
+        index = new_dd_object(dsk, index_name, pd.RangeIndex(0, 1), divisions)
+        arrays_and_indices.extend([index.__dask_layers__()[0], "i"])
+        graph_dependencies.append(index)
+        numblocks[index.__dask_layers__()[0]] = (index.npartitions,)
+
+    blk = blockwise(
```

Comment:
Is using `blockwise` directly necessary, or could `map_partitions` work? `map_partitions` is just a little easier to read IMO.

Reply:
I don't have super strong priors here, but to me the contract for `map_partitions` is that everything should be dataframe-ish, so mixing arrays and dataframes feels fragile (even if it works today).

I'd rather spend effort making `blockwise` easier to get right than forcing stuff through `map_partitions` that doesn't quite fit.

Reply:
My opinion is that the contract for `map_partitions` (the user-facing function) should support mixing arrays and dataframes (and bags?)—that's a very useful thing to be able to do as a user (and it already works).

If `index` ends up being an actual `Index` object as I suggested, then I think using `map_partitions` is perfectly appropriate here—all the inputs would just be Arrays or DataFrame-y things, which is what `map_partitions` is supposed to handle.

I'd also like to make `blockwise` easier to get right, but it's a different interface (at the graph level as opposed to the higher collection level). If it's possible to express the whole thing at the collection level, I feel like we might as well.

Reply:
DataFrame ACA temporarily [converts to bag](https://github.com/dask/dask/blob/main/dask/dataframe/core.py#L6070-L6079) in cases where the partitions may not be frame-like. It seems that dataframe, array, and bag are all pretty much the same, but bag doesn't include metadata compliance in the "contract".  I suppose we could make it easier to temporarily cast an array collection to bag as well.